library/objective-c/EnvoyEngineImpl.m

```diff
@@ -43,4 +49,43 @@ - (int)runWithConfigYAML:(NSString *)configYAML logLevel:(NSString *)logLevel {
                                            callbacks:callbacks];
 }
 
+#pragma mark - Reachability
+
+static SCNetworkReachabilityRef _reachability_ref;
+
+static void _reachability_callback(SCNetworkReachabilityRef target,
+                                   SCNetworkReachabilityFlags flags, void *info) {
+  if (flags == 0) {
+    return;
+  }
+
+  BOOL isUsingWWAN = flags & kSCNetworkReachabilityFlagsIsWWAN;
+  set_preferred_network(isUsingWWAN ? ENVOY_NET_WWAN : ENVOY_NET_WLAN);
+}
+
+static void _ensure_reachability_running() {
+  if (_reachability_ref) {
+    return;
+  }
+
+  NSString *name = @"io.envoymobile.reachability";
+  SCNetworkReachabilityRef reachability =
+      SCNetworkReachabilityCreateWithName(nil, [name UTF8String]);
+  if (!reachability) {
+    return;
+  }
+
+  _reachability_ref = reachability;
+
+  SCNetworkReachabilityContext context = {0, NULL, NULL, NULL, NULL};
+  if (!SCNetworkReachabilitySetCallback(_reachability_ref, _reachability_callback, &context)) {
+    return;
+  }
+
+  dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
```

Comment:
Thinking we should maybe use a dedicated queue here since the main thread may get blocked on slow render operations and also be initiating network requests.

Reply:
Oh sorry, you are using separate queue from the main one. What do you think about making this a higher priority?

Reply:
We can create our own queue if you think that makes sense. It is worth noting that this is not the main thread though, this is a background queue with default priority

Reply:
We could use `DISPATCH_QUEUE_PRIORITY_HIGH`, sure